internal/twirptest/client_test.go

```diff
@@ -257,8 +257,94 @@ func TestClientRedirectError(t *testing.T) {
 	})
 }
 
+func TestClientWithHooks(t *testing.T) {
+	requestPreparedCalled := false
+	responseReceivedCalled := false
+	errorCalled := false
+
+	hooks := &twirp.ClientHooks{
+		RequestPrepared: func(ctx context.Context, req *http.Request) (context.Context, error) {
+			requestPreparedCalled = true
+			return ctx, nil
+		},
+		ResponseReceived: func(ctx context.Context) {
+			responseReceivedCalled = true
+		},
+		Error: func(ctx context.Context, err twirp.Error) {
+			errorCalled = true
+		},
+	}
+
+	h := PickyHatmaker(1)
+	s := httptest.NewServer(NewHaberdasherServer(h, nil))
+	defer s.Close()
+
+	// Clients
+	protoCli := NewHaberdasherProtobufClient(s.URL, &http.Client{}, twirp.WithClientHooks(hooks))
+	jsonCli := NewHaberdasherJSONClient(s.URL, &http.Client{}, twirp.WithClientHooks(hooks))
+	ctx := context.Background()
+	var resp *Hat
+	var err error
+
+	// Test proto success
+	resp, err = protoCli.MakeHat(ctx, &Size{Inches: 1})
+	if err != nil {
+		t.Fatalf("Proto client method returned unexpected error: %s", err)
+	}
+	if resp == nil {
+		t.Fatalf("Proto client method expected to return non-nil response, but it is nil")
+	}
+
+	if requestPreparedCalled == false || responseReceivedCalled == false {
+		t.Fatalf("expected both requestPreparedCalled and responseReceivedCalled to be true")
+	}
+
+	// Test proto failure
+	resp, err = protoCli.MakeHat(ctx, &Size{Inches: 666})
+	if err == nil {
+		t.Fatalf("Proto client method expected to fail, but error is nil")
+	}
+	if resp != nil {
+		t.Fatalf("Proto client method expected to return nil response on error, but returned non-nil")
+	}
+
+	if errorCalled == false {
```

Comment:
I think you need to reset the test state before this check. That might mean a refactor of this test in general.